### PR TITLE
Fixed UI inputs caret position and focus

### PR DESCRIPTION
**Pull request by xGuTeK**
character deletion works and sets the text append correctly

This PR resolves: https://github.com/stevewgr/ko/issues/51

### CODE REVIEW

File: src/engine/N3Base/N3UIEdit.cpp [hunk 1]
```diff
@@ -752,10 +758,22 @@ void CN3UIEdit::UpdateCaretPosFromEditCtrl()
...
+	}
+	else {
+		std::string getText = s_pFocusedEdit->GetString();
+		cursorPosition = getText.size();
+
+		SendMessageA(s_hWndEdit, EM_SETSEL, (WPARAM)cursorPosition, (LPARAM)cursorPosition);
+	}
+
+
+	s_pFocusedEdit->SetCaretPos(cursorPosition);
```
Comment (stevewgr) [inline src/engine/N3Base/N3UIEdit.cpp:776]: We could write the same thing like this:
```suggestion
	int cursorPosition = 0;
	if (changeFocus) {
		cursorPosition = s_pFocusedEdit->GetString().size();
		SendMessage(s_hWndEdit, EM_SETSEL, (WPARAM)cursorPosition, (LPARAM)cursorPosition);
	} else {
		cursorPosition = LOWORD(SendMessage(s_hWndEdit, EM_GETSEL, NULL, NULL));
	}

	s_pFocusedEdit->SetCaretPos(cursorPosition);
```

I'm not sure however if `EM_SETSEL` is the right pick here. I would have to check further to tell.

File: src/engine/N3Base/N3UIEdit.h [hunk 2]
```diff
@@ -55,7 +55,7 @@ class CN3UIEdit : public CN3UIStatic
 	static BOOL				CreateEditWindow(HWND hParent, RECT rect);
 	static LRESULT APIENTRY EditWndProc(HWND hWnd, WORD Message, WPARAM wParam, LPARAM lParam);
 	static void				UpdateTextFromEditCtrl();
-	static void				UpdateCaretPosFromEditCtrl();
+	static void				UpdateCaretPosFromEditCtrl(bool changeFocus);
```
Comment (stevewgr) [inline src/engine/N3Base/N3UIEdit.h:58]: I would rename `changeFocus` to `bNeedsFocus`.

File: src/game/Main.cpp [hunk 3]
```diff
@@ -44,7 +44,15 @@ LRESULT CALLBACK WndProcMain(HWND hWnd, UINT message, WPARAM wParam, LPARAM lPar
...
+					if (GetAsyncKeyState(VK_TAB) || GetAsyncKeyState(VK_LBUTTON)) {) {
+						pEdit->UpdateCaretPosFromEditCtrl(true);
+					}
+					else {
+						pEdit->UpdateCaretPosFromEditCtrl(false);
+					}
+
+
+					
 					CGameProcedure::SetGameCursor(CGameProcedure::s_hCursorNormal);
```
Comment (stevewgr) [inline src/game/Main.cpp:56]: There is syntax error here and it fails to compile: `{)`. We can also write it this way to improve readability:
```suggestion
					pEdit->UpdateTextFromEditCtrl();
					bool bNeedsFocus = GetAsyncKeyState(VK_TAB) || GetAsyncKeyState(VK_LBUTTON);
					pEdit->UpdateCaretPosFromEditCtrl(bNeedsFocus);
					CGameProcedure::SetGameCursor(CGameProcedure::s_hCursorNormal);
```

**Review by stevewgr — CHANGES_REQUESTED** [review]
Hi @xGuTeK, thanks for this great PR!
I left couple of comments in code.
There is another issue I noticed about when the window is minimized or ALT-TABBED and then focused again, the edit box loses focus. I think it's best when login procedure is active, to set the inputs focus. Also listen to `WM_ACTIVATE` events in `WndProcMain` and setting up the focus. Otherwise one must click or tab in order to get the focus and it's not ideal.

One last thing: if you could please squash your commits, that would be great. Since the changes in your own fork, after you squash the commits you can `git push -f` force push.

Quite tricky thing to implement, since it touches couple of pieces here. Great work man and thanks a lot for the contributions.

File: src/engine/N3Base/N3UIEdit.cpp (hunk 1 above)
Comment (xGuTeK) [reply src/engine/N3Base/N3UIEdit.cpp:776]: I that function there was variable int iLen = GetWindowTextLength(s_hWndEdit); too get editbox length but this variable is never used  and it was commented out

**Pull request closed without merging**